### PR TITLE
A refined solution to the beta-iota discrepancies between 8.4 and 8.5/8.6 "refine"

### PR DESCRIPTION
**Pull request by herbelin**
There is a long story of commits trying to improve the compatibility between 8.4 and 8.5 refine, as discussed in [#346](https://github.com/coq/coq/pull/346).

The current commit tries to identify (one of?) the exact points of divergence between 8.4 and 8.5 refine, namely the beta-iota normalisation done in the types of arguments of constructors in pattern-matching.

For the record, note that for the conclusion of each new goal, there were a nf_betaiota in 8.4 done in function Evarutil.evars_to_metas, so the compatibility expects that such a nf_betaiota on the conclusion of each goal remains.

See also bug report [#5219](https://coq.inria.fr/bugs/show_bug.cgi?id=5219) for elements of discussion.

### CODE REVIEW

**Comment by mattam82** [comment]
Shouldn't this code be subject to a compatibility flag? Also, what is the rationale for beta-iota normalizing these types, what are the examples? I would guess if you have higher-order parameters or something like that?

**Comment by silene** [comment]
There is something I don't understand. As far as I can tell from Jason's testcase, both 8.4 and 8.5 work correctly, only 8.6 is broken. Yet you seem to imply that both 8.5 and 8.6 are broken.

**Comment by herbelin** [comment]
@silene: Yes, only 8.6 is "broken" for the example in #5219, but if you look at the discussion in #346,
you can see that if it works in 8.5, it is because of a hack which produces more harm than what it is supposed to solve.

Indeed the hack in 8.5 is to normalize _all_ hypotheses of the goal just to have the same normalizations that 8.4 makes only in _some_ of the hypotheses _generated_ by `refine`: as far as I could see:
- 8.4's `refine` does not normalize the hypotheses corresponding to "fun" (it drops the explicit type of a "fun" and inherits the type given as a constraint),
- 8.4's `refine` does not normalize the hypotheses originally present,
- 8.4's `refine` normalizes only the types of the variables coming from a pattern-matching problem.

So, in short, 8.5's `refine` is broken because it normalizes _everywhere_. 8.6's `refine` is broken wrt 8.4's `refine` because it does not normalize the type of hypotheses created by pattern-matching.

@mattam82: Yes, the example is exactly the one given in [#5219](https://coq.inria.fr/bugs/show_bug.cgi?id=5219):

```coq
Goal forall x : sigT (fun x => x = 1), True.
  intro x; refine match x with existT _ x' e' => _ end.
```
where `e'` is given the type `(fun x => x = 1) x'` by the pattern-matching compilation algorithm, while in 8.4, it was using the prim_refiner which was giving the type `x'=1` to `e'`.

The original rationale is to simulate 8.4's `refine`. But I think it is anyway better to beta-iota normalize, even though, towards a more ideal world, my recommendation would be not to go to nf but only to surgically whd normalize at the time of substituting each higher order parameter or each constructor (as in `meta_reducible_instance`), so as to also preserve original redexes intentionally introduced by the user.

Yes, we can add a compatibility flag if we want (I'm currently testing on contribs to evaluate if it is worth).

**Comment by silene** [comment]
If the behavior of 8.6 is meant to be indistinguishable from the one of 8.5 (with respect to normalization of case branches), I don't see much point in having a compatibility flag. Am I misunderstanding something?

**Comment by herbelin** [comment]
@silene: Yes, it would be indistinguishable from 8.4 and 8.5 behavior of `refine`, but it might have some other consequences elsewhere, e.g. for type inference since at the time of typechecking say
`fun x : sigT (fun x => x = 1) => match x with existT _ x' e' => some-term end` the subterm `some-term` will be in a context `e':x'=1` instead of `e':(fun x => x = 1) x'` as it is now.

This is where a compatibility flag could virtually be needed...

But I tend to think that the new type is more desirable anyway in general (up to the former remark that a surgical beta-iota would be better than a full beta-iota).

**Comment by herbelin** [comment]
Actually, there would be some fun for a specialist of abstract machines in implementing a surgical beta-iota, as it would require to recognize on the fly the evaluation context of the variables to which the substitution is applied!

**Comment by herbelin** [comment]
Any objection thus to merging this?

(And I'm more and more thinking that the idea of a "surgical beta-iota" should be explored!)

**Comment by maximedenes** [comment]
This a compatibility breaking change, right? If yes, I object to its merge until we discuss the compatibility policy at the working group.

**Comment by herbelin** [comment]
This is strictly speaking a compatibility breaking change.

It has been tested on contribs without leading to failures (see https://ci.inria.fr/coq/view/coq-contribs/job/coq-contribs/604/).

**Comment by maximedenes** [comment]
Yes, but it has happened many times before that we didn't detect compatibility issues on the contribs, while on other developments like Bedrock, HoTT, Fiat an others, it might be an issues for users. Note that I'm not opposing this change (I don't feel qualified to evaluate it), I simply would like that we discuss it at the next working group before merging.

**Comment by herbelin** [comment]
@maximedenes: What are the last news? Do you plan to discuss a "compatibility policy" and this PR at the next WG, or can we simply consider it to be a "risk" for improvement worthwhile to be made?

**Comment by Zimmi48** [comment]
@herbelin Now there is Travis to test on more recent developments. You can trigger the build on these by rebasing or closing/reopening the PR (I think).

**Review by aspiwack — APPROVED** [review]
As @herbelin says, it may be a bit overzealous a reduction, but it's in a rather restricted situation, so we shouldn't have many surprises.

I don't know what a good, robust, heuristic would be for post-refine reductions. This patch has the benefit of being very localised, and orthogonal to everything else, so it won't damage further improvement (or even revert). So I don't see a reason for not merging.

**Comment by maximedenes** [comment]
As I explained during the last WG, the history of those changes in reductions done in `refine` is very complex, but the compatibility of this PR doesn't seem too bad, so let's merge it.

**Comment by maximedenes** [comment]
@herbelin Could you add an entry in `CHANGES`? Thanks!

**Comment by herbelin** [comment]
@maximedenes: this patch is not about `refine` properly speaking. The tactic `refine` was reimplemented in 8.5 and it is clear that the compatibility is not strict. Previously `refine` was relying on `logic.ml` for typing, now it relies on `pretyping.ml` and they may infer different types. 

So, I think it is an error to look at this PR as a PR only about `refine`. It is a PR which proposes to change the result of the type of the pattern variables in `match`, as in the example:
```coq
Check match x with existT _ x' e' => _ end.
(* (let (x'0, e'0) := x in ?t@{x'0:=x'; e'0:=e'; x':=x'0; e':=e'0}) : True
     : True
where
?t : [x : {x : nat & x = 1}   x'0 : nat   e'0 : (fun x : nat => x = 1) x'0   x' : nat   e' : (fun x : nat => x = 1) x' |- True]  *)
```
where you can see that the type of the variables `e'` and `e'0` given by `pretyping.ml` are not β-reduced, while for a `match` typed by `logic.ml` they would have been β-reduced.

So, independently of `refine`, do we agree that it seems more natural, more expected, to have the type of `e'` and `e'0` β-reduced, or _at least_ to implement the surgical β-reduction I already mentioned a couple of time?

So, yes, the history of changes made for reduction in `refine` is complex. We made repeated mistakes and the reason is that we did not realize from the very beginning that the real source of the problem was indeed pretty indirect.

So, yes, this is yet another instance of the (hyper)sensitivity of Coq on the way key components such as unification and type inference are implemented, and I have no better solution to it except to maintain for some time old versions of these components for compatibility and to _concentrate our efforts_ on what we consider are better, simpler, more regular, implementations of these components, as @mattam82 proposed.

In particular, it might make sense as soon as 8.7 to introduce this versioning of the key components of Coq. Then, this PR could be specifically attached to version 8.7 and deactivated in the 8.X (X<7) compatibility modes, if the consensus is to go this way. What do you think?

About `CHANGES`, I'm unsure this file is appropriate for the level of technicality of the change, but I can try. Would you be ok with a sentence such as "The types of variables in patterns of `match` are now beta-iota-reduced after type-checking.  This has an impact on the type of the variables that the tactic `refine` introduces in the context, producing types a priori closer to the expectations."?

**Comment by mattam82** [comment]
I'm ok with merging that change too, and I agree it is more natural indeed to do the beta-iota reduction in types of match variables. Your sentence for CHANGES looks good to me as well. I'm realizing we're missing a "refinement" chapter in the reference manual where we could document these choices and hopefully one day unification.

**Comment by herbelin** [comment]
BTW: what are the plans wrt unification? It happens that some of us accumulated some knowledge about unification and I wonder how to make this useful towards the objective of improving and documenting unification.

**Comment by maximedenes** [comment]
@herbelin Thanks for the clarification! And yes, I think the sentence in `CHANGES` is fine. Can you add it? I'll merge this PR right after. Thanks!

As for unification, it would be good to find some time to work on a plan with Matthieu. Maybe at the implementors workshop?

**Pull request merged**
